app/controllers/summaries_controller.rb

```diff
@@ -15,7 +16,42 @@ def category_totals
     end
   end
 
-  def sum_purchases_in_category(cat)
-    Purchase.where(subcategory: cat.subcategories).to_a.map(&:amount).sum
+  def sum_purchases_in_category(category)
+    Purchase.where(subcategory: category.subcategories).to_a.map(&:amount).sum
+  end
+
+  def month_totals
+    first_purchase = Purchase.all.order(:date).first[:date]
+    last_purchase = Purchase.all.order(:date).last[:date] # this could be Date.today rather than most recent purchase
+    date_range = (first_purchase..last_purchase).map { |d| d.strftime('%B %Y') }.uniq
+    sum_month(category_totals_per_month.flatten.group_by { |n| n[:name] }, date_range)
+  end
+
+  def category_totals_per_month
+    Category.all.map do |c|
+      sum_purchases_by_month(c).map do |p|
+        {
+          name: c.name,
+          date: p[:date],
+          amount: p[:amount]
+        }
+      end
+    end
+  end
+
+  def sum_purchases_by_month(category)
+    Purchase.where(subcategory: category.subcategories).map { |p| { date: p.date, amount: p.amount } }
+  end
+
+  def sum_month(month, date_range)
+    month.map { |key, value| { name: key, data: value.group_by_month { |n| n[:date] }.map { |key, value| [key.strftime('%B %Y'), value.map { |v| v[:amount] }.sum] } } }.each do |c|
+      contained_months = c[:data].map(&:first)
+      blank_months = date_range - contained_months
+      blank_months.each { |b| insert_empty_month(c[:data], [b, 0]) }
+    end
```

Comment:
This is pretty dense and to be honest I haven't got to grips with it totally, it will be easier once the other naming issues are sorted out. Just wondering if it's possible to replace this FN with something like this (not proper code):
```ruby
def sum_per_category_and_month(purchases_by_category, all_months)
    purchases_by_category.map do |cat|
      summed = # some code
      # Result like:
      #  { name: 'Basics', data: [['November 2018', 100], ['December 2018', 150]] }

      zero_purchase_months = # months that are in all_months, but not summer.data
      # Add in months with total of zero:
      ?zero_purchase_months.each { |m| summed.data.push({[m, 0]}) }

      summed
    end
  end
```

Some of that is similar to what you have already, I'm just thinking it through in the writing.

Reply:
What the existing method offers which this alternative doesn't is maintaining month order - will take another look at refactoring once other concerns are addressed.